homeassistant/components/onvif/__init__.py

```diff
@@ -113,6 +122,39 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
     )
 
 
+async def _get_snapshot_auth(hass, device, entry):
+    if not (device.username and device.password):
+        return HTTP_DIGEST_AUTHENTICATION
+
+    snapshot_uri = await device.async_get_snapshot_uri(device.profiles[0])
+    auth = HTTPDigestAuth(device.username, device.password)
+
+    def _get():
+        # so we can handle keyword arguments
+        return requests.get(snapshot_uri, timeout=1, auth=auth)
```

Comment:
We don't want transport code in the integration. Can we use a library command to get similar behavior?

Reply:
I'm using the same code that's being used to grab the snapshot (https://github.com/home-assistant/core/pull/38729/files#diff-dfca6ffe288b877cb7915d2dc1e6d13fR132),  I'll defer to @hunterjm on this.

Reply:
@MartinHjelmare - We've talked about this before. This isn't really a request to get API data (which should absolutely be wrapped), but to fetch an image. `requests` or `aiohttp` are used directly to do so in almost every camera implementation.

Generic Camera: https://github.com/home-assistant/core/blob/dev/homeassistant/components/generic/camera.py#L132-L165
MJPEG Camera: https://github.com/home-assistant/core/blob/dev/homeassistant/components/mjpeg/camera.py#L132-L145
Netatmo: https://github.com/home-assistant/core/blob/dev/homeassistant/components/netatmo/camera.py#L181-L190
and so on...

I'm not sure what the best way to address this would be.

Reply:
All those examples are breaking our standard. We shouldn't extend that. We're fetching data (an image) from the device. That should be handled by the device library.

Reply:
Can we fix this in a followup PR and merge this one? I'll look into the library and change in both places. Since this is categorized as a feature, rather than a bug, and without it the ONVIF integration is unusable in some cases, I would love to get it in before the beta.
Thanks.

Reply:
Ok, I'll approve this. Ping me when you have a refactor with extraction of the transport code. Thanks!